Mapsios 247 sync location examples with android

pjleonard37 opened the pull request:
<!--
Thanks for submitting a pull request!

Please fill out the sections below to complete your submission.

We appreciate your contributions!
-->

<!--
Describe the changes in this PR here.

• If this is a new feature, include a short summary on how to use it.
• If this is a bug fix, explain how your contribution resolves the problem.
• Include before/after visuals or gifs if this PR includes visual changes.
• Add a line with "Fixes: #issue-number" or "Fixes: issue URL" for each publicly-visible issue that is fixed by this PR.
-->
This PR aligns our puck customization example with the Android example found here: https://github.com/mapbox/mapbox-maps-android/blob/main/app/src/main/java/com/mapbox/maps/testapp/examples/LocationComponentActivity.kt 

Android: 
<img src="https://user-images.githubusercontent.com/7976026/187959706-62031357-e7ec-41a2-8792-9bef89868ef1.png" width="200" />

iOS:


https://user-images.githubusercontent.com/7976026/187965769-c0fb2285-abf2-4da1-9be0-2adc37656b88.mp4




## Pull request checklist:
 - [X] Describe the changes in this PR, especially public API changes.
 - [X] Include before/after visuals or gifs if this PR includes visual changes.
    <!--
        | Before | After |
        | ----- | ----- |
        | <img src="" width = 250/> | <img src="" width = 250/> |
        or
        | <video src="" width = 250/> | <video src="" width = 250/> |
    -->

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-ios/blob/main/CONTRIBUTING.md#contributor-license-agreement).

OdNairy commented on Apps/Examples/Examples/All Examples/CustomPuckExample.swift:
```diff
@@ -0,0 +1,160 @@
+        button.backgroundColor = #colorLiteral(red: 0, green: 0.4784313725, blue: 0.9882352941, alpha: 1)
+        button.translatesAutoresizingMaskIntoConstraints = false
+        button.layer.cornerRadius = 20
+        button.clipsToBounds = true
```
Why do we need this?

OdNairy commented on Apps/Examples/Examples/All Examples/CustomPuckExample.swift:
```diff
@@ -0,0 +1,160 @@
+        view.addSubview(button)
+
+        // Set button location
+        let horizontalConstraint = button.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -24)
```
It is vertical

OdNairy commented on Apps/Examples/Examples/All Examples/CustomPuckExample.swift:
```diff
@@ -0,0 +1,160 @@
+
+        // Set button location
+        let horizontalConstraint = button.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -24)
+        let verticalConstraint = button.centerXAnchor.constraint(equalTo: view.centerXAnchor)
```
And it is horizontal

OdNairy commented on Apps/Examples/Examples/All Examples/NavigationSimulatorExample.swift:
```diff
@@ -41,7 +41,8 @@ final class NavigationSimulatorExample: UIViewController, ExampleProtocol {
         navigationSimulator = NavigationSimulator(viewport: mapView.viewport, route: sampleRouteLine)
 
-        mapView.location.options.puckType = .puck2D(.makeDefault(showBearing: true))
+        let configuration = Puck2DConfiguration(topImage: UIImage(named: "user_puck_icon"))
```
It is important to have an image in this place. We have to unwrap it so in case of missing image we would have a crash with exact place in code describing the issue.
```suggestion
        let configuration = Puck2DConfiguration(topImage: UIImage(named: "user_puck_icon")!)
```

OdNairy commented on Apps/Examples/Examples/Assets.xcassets/user_puck_icon.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+{
```
Want to confirm that missing "preserving vector data" was intentional with SVG image.

pjleonard37 replied:
It was not intentional -- I updated to include

OdNairy reviewed (changes requested):
Please, use a separate enum for toggles. Do not manipulate by UI data to check the model state.

pjleonard37 commented on Apps/Examples/Examples/All Examples/CustomPuckExample.swift:
```diff
@@ -0,0 +1,160 @@
+                                      preferredStyle: .actionSheet)
+
+        alert.addAction(UIAlertAction(title: "Toggle Puck visibility", style: .default, handler: { [weak self] _ in
+            self!.toggePuckVisibility()
```
Complete in update

pjleonard37 commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -1,86 +0,0 @@
-import MapboxMaps
-
-@objc(Custom2DPuckExample)
-public class Custom2DPuckExample: UIViewController, ExampleProtocol {
```
Yes: updated

pjleonard37 commented:
I updated the example in response to @OdNairy's above. The most major change is creating a separate enum for each toggle to handle model state. When they are updated the UI changes follow from that state.

maios commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -4,27 +4,125 @@ import MapboxMaps
     internal var mapView: MapView!
-    internal var showsAccuracyRing: Bool = false {
+    internal var puckConfiguration = Puck2DConfiguration.makeDefault(showBearing: true)
+    private var showsPuck: PuckVisibility = .isVisible {
```
imo this naming fits our convention better. But please feel free to dismiss this suggestion :-D
```suggestion
    private var isPuckVisible: PuckVisibility = .isVisible {
```

OdNairy replied:
The `showsPuck` name sounds good for me.

maios commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -4,27 +4,125 @@ import MapboxMaps
+        }
+    }
+
+    private enum PuckVisibility {
```
is there a reason why are using a custom enum here instead of a boolean?

OdNairy replied:
I fully support custom enums over boolean types due to better readability in code.

maios commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -4,27 +4,125 @@ import MapboxMaps
+        didSet {
+            puckConfiguration.topImage = puckImage == .blueDot ? .none : UIImage(named: "star")
+            // Only add puck with new image to the map if it is meant to be visible
+            if showsPuck == .isVisible {
```
this snippet has been duplicated in many places, could we find a way to have a single implementation for updating puck configuration upon changes?

pjleonard37 replied:
Yes -- can do 👍

maios commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -50,37 +142,72 @@ public class Custom2DPuckExample: UIViewController, ExampleProtocol {
+        }))
+
+        alert.addAction(UIAlertAction(title: "Toggle Projection", style: .default, handler: { [weak self] _ in
+            self?.projection.toggle()
```
we don't need a weak reference of `self` inside alert action's handler, given that we don't retain the alert controller.

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -4,27 +4,125 @@ import MapboxMaps
+
+    private var showsAccuracyRing: PuckAccuracyRingVisibility = .isHidden {
+        didSet {
+            puckConfiguration.showsAccuracyRing = showsAccuracyRing == .isVisible ? true : false
```
```suggestion
            puckConfiguration.showsAccuracyRing = showsAccuracyRing == .isVisible
```

OdNairy replied:
Even better solution would be to add `var isVisible: Bool` to the `PuckVisibility` enum and use it as:
```suggestion
            puckConfiguration.showsAccuracyRing = showsAccuracyRing.isVisible
```

OdNairy commented on Apps/Examples/Examples/All Examples/NavigationSimulatorExample.swift:
```diff
@@ -41,7 +41,8 @@ final class NavigationSimulatorExample: UIViewController, ExampleProtocol {
         navigationSimulator = NavigationSimulator(viewport: mapView.viewport, route: sampleRouteLine)
 
-        mapView.location.options.puckType = .puck2D(.makeDefault(showBearing: true))
+        let configuration = Puck2DConfiguration(topImage: UIImage(named: "user_puck_icon")!)
+        mapView.location.options.puckType = .puck2D(configuration)
```
Double-check that this change was intentional

pjleonard37 replied:
Yes: to match the Android puck

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -50,37 +142,72 @@ public class Custom2DPuckExample: UIViewController, ExampleProtocol {
 
-    @objc func showHideAccuracyRadius() {
-        showsAccuracyRing.toggle()
+    private func customizePuckButton() {
```
The name of `customizePuckButton` sounds like it will return `UIButton` but it doesn't.
```suggestion
    private func addCustomizePuckButton() {
```

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -50,37 +142,72 @@ public class Custom2DPuckExample: UIViewController, ExampleProtocol {
+        button.setTitle("Customize puck", for: .normal)
+        button.setTitleColor(.white, for: .normal)
+        button.backgroundColor = #colorLiteral(red: 0, green: 0.4784313725, blue: 0.9882352941, alpha: 1)
+        button.translatesAutoresizingMaskIntoConstraints = false
```
It is a good practice to keep `translatesAutoresizingMaskIntoConstraints = false` close to the constraints setup. It would help you not to forget this part on next views.

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -50,37 +142,72 @@ public class Custom2DPuckExample: UIViewController, ExampleProtocol {
+        view.addSubview(button)
+
+        // Set button location
+        let horizontalConstraint = button.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -24)
+        let verticalConstraint = button.centerXAnchor.constraint(equalTo: view.centerXAnchor)
+        let widthConstraint = button.widthAnchor.constraint(equalToConstant: 200)
+        let heightConstraint = button.heightAnchor.constraint(equalToConstant: 40)
+        NSLayoutConstraint.activate([horizontalConstraint, verticalConstraint, widthConstraint, heightConstraint])
```
Nit: There is no need to make a variables if you are not going to change them later
```suggestion
        NSLayoutConstraint.activate([
            button.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -24),
            button.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            button.widthAnchor.constraint(equalToConstant: 200),
            button.heightAnchor.constraint(equalToConstant: 40)
        ])
```

OdNairy reviewed (changes requested):
This version looks much better, well done
There is the same pattern multiple times in the code – you get a `didSet` callback, check the enum, and based on this ternary condition you assign the value to the corresponding field in puckConfiguration. 
Since Swift allows custom enum functions, it would be better to put such kind of ternary conditions (and effectively convert them into switch-case statement) as a computed variable. That way you can do the things like `puckConfiguration.topImaeg = puckImage.image` without extra logic in Presentation layer.
It also make sense to try to extract all puck UI related `didSet` callback into the single method (something like `updatePuckUI`) and call it from `didSet` callbacks. The overhead would be very little, but you would aggregate all UI-related changes in the single place

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -4,27 +4,125 @@ import MapboxMaps
+    private var showsBearing: PuckBearingVisibility = .isVisible {
+        didSet {
+            // It's necessary to create a new default when switching bearing visibility
+            var newPuck = showsBearing == .isVisible ? Puck2DConfiguration.makeDefault(showBearing: true) : Puck2DConfiguration.makeDefault(showBearing: false)
```
I would like you to rewrite this whole approach with `didSet`'s, but this particular thing if your implementation had to be the following:
```suggestion
            var newPuck = Puck2DConfiguration.makeDefault(showBearing: showsBearing == .isVisible) 
```

pjleonard37 commented:
Thanks @OdNairy and @maios for the recommendations -- I've incorporated them into the latest commit. I moved all of the logic I could into the enums, separated out an `updatePuckUI()` function, and accepted the smaller recommendations.

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -4,27 +4,158 @@ import MapboxMaps
+
+    private var puckImage: PuckImage = .blueDot {
+        didSet {
+            puckConfiguration.topImage = puckImage.image
+            updatePuckUI()
```
It would be something similar to SwiftUI where you have a state variables and the single place to update the UI representation.

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -4,27 +4,158 @@ import MapboxMaps
+
+    private var projection: StyleProjectionName = .mercator {
+        didSet {
+            do {
+                try mapView.mapboxMap.style.setProjection(StyleProjection(name: projection))
+            } catch {
+                print(error)
+            }
```
Can you extract this to the separate function as well? We usually use very lightweight didSet observers.

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -50,37 +175,79 @@ public class Custom2DPuckExample: UIViewController, ExampleProtocol {
+                                      message: "Select an options to toggle.",
+                                      preferredStyle: .actionSheet)
+
+        alert.addAction(UIAlertAction(title: "Toggle Puck visibility", style: .default, handler: { _ in
+            self.showsPuck.toggle()
+        }))
```
Nit: We do usually prefer the special syntax for trailing closure
```suggestion
        alert.addAction(UIAlertAction(title: "Toggle Puck visibility", style: .default) { _ in
            self.showsPuck.toggle()
        })
```

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -50,37 +175,79 @@ public class Custom2DPuckExample: UIViewController, ExampleProtocol {
+    }
+
+    func updatePuckUI() {
+        if showsPuck == .isVisible {
+            mapView.location.options.puckType = .puck2D(puckConfiguration)
+        }
```
Should we hide puck here if it is not visible?

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -50,37 +175,79 @@ public class Custom2DPuckExample: UIViewController, ExampleProtocol {
+    }
+
+    func updatePuckUI() {
+        if showsPuck == .isVisible {
```
An idea behind enum-per-state is that you then use them in the `swift-case` statement. It would make a compilation time checker over any new type of state (`case`).

pjleonard37 commented:
Thanks for the additional review, @OdNairy. I've shifted the logic further to ensure that the `didSet`s are lightweight and proper syntax is used.

OdNairy commented on Apps/Examples/Examples/All Examples/Custom2DPuckExample.swift:
```diff
@@ -50,37 +169,93 @@ public class Custom2DPuckExample: UIViewController, ExampleProtocol {
+    func updatePuckUI() {
+        puckConfiguration = Puck2DConfiguration.makeDefault(showBearing: showsBearing.isVisible)
+        puckConfiguration.showsAccuracyRing = showsAccuracyRing.isVisible
+        puckConfiguration.topImage = puckImage.image
+        switch showsPuck {
```
nit
```suggestion
        puckConfiguration.topImage = puckImage.image
        
        switch showsPuck {
```

OdNairy reviewed (approved):
LGTM! 🎉
